app/src/main/java/org/p2p/wallet/home/ui/main/UserTokensPolling.kt

```diff
@@ -19,37 +29,68 @@ class UserTokensPolling(
     private val ethAddressEnabledFeatureToggle: EthAddressEnabledFeatureToggle,
     private val ethereumInteractor: EthereumInteractor,
-) {
+    private val ethTokenProvider: EthereumTokensProvider,
+    private val dispatchers: CoroutineDispatchers,
+) : CoroutineScope {
+    override val coroutineContext: CoroutineContext
+        get() = SupervisorJob() + dispatchers.io
 
     private val isPollingEnabled: Boolean
         get() = appFeatureFlags.isPollingEnabled.featureValue
 
-    suspend fun startPolling(onTokensLoaded: suspend (List<Token.Active>, EthereumHomeState) -> Unit) {
-        if (isPollingEnabled) {
-            try {
-                while (true) {
-                    delay(POLLING_DELAY.inWholeMilliseconds)
-                    val newTokens = userInteractor.loadUserTokensAndUpdateLocal()
-                    val ethereumState = getEthereumState()
-                    onTokensLoaded.invoke(newTokens, ethereumState)
-                    Timber.d("Successfully auto-updated loaded tokens")
+    private val solTokensFlow = MutableStateFlow<List<Token.Active>>(emptyList())
+
+    fun shareTokenPollFlowIn(scope: CoroutineScope): StateFlow<Pair<List<Token.Active>, List<Token.Eth>>> =
+        solTokensFlow.combine(ethTokenProvider.erc20Tokens) { sol, eth ->
+            sol to eth
+        }.stateIn(scope, SharingStarted.WhileSubscribed(), Pair(emptyList(), emptyList()))
+
+    fun startPolling() {
+        startPollingSolana()
+        startPollingEthereum()
+    }
+
+    suspend fun refresh() {
+        solTokensFlow.emit(emptyList())
+    }
+
+    private fun startPollingSolana() {
+        launch {
+            if (isPollingEnabled) {
+                try {
+                    while (true) {
+                        delay(POLLING_DELAY.inWholeMilliseconds)
+                        val newTokens = userInteractor.loadUserTokensAndUpdateLocal()
+                        solTokensFlow.emit(newTokens)
+                        Timber.d("Successfully auto-updated loaded tokens")
+                    }
+                } catch (e: CancellationException) {
+                    Timber.i("Cancelled tokens remote update")
+                } catch (e: Throwable) {
+                    Timber.e(e, "Failed polling tokens")
                 }
-            } catch (e: CancellationException) {
-                Timber.i("Cancelled tokens remote update")
-            } catch (e: Throwable) {
-                Timber.e(e, "Failed polling tokens")
+            } else {
+                Timber.d("Skipping tokens auto-update")
             }
-        } else {
-            Timber.d("Skipping tokens auto-update")
         }
     }
 
-    private suspend fun getEthereumState(): EthereumHomeState {
-        if (!ethAddressEnabledFeatureToggle.isFeatureEnabled) {
-            return EthereumHomeState()
+    private fun startPollingEthereum() {
+        launch {
+            if (ethAddressEnabledFeatureToggle.isFeatureEnabled) {
```

Comment:
you can reduce indentation moving check for feature toggle in the beginning of the method

if (!ethAddressEnabledFeatureToggle.isFeatureEnabled){
    return 
}